Update default theme code highlighting with Atom's One Light colors, consolidate theme variables

mattxwang opened the pull request:
This PR replaces the default light code highlighting theme with Atom's One Light theme colors. This should provide more visual similarity with our dark theme, and as a byproduct, fix some of the contrast issues from our current light theme.

In addition (different from the original purpose of this PR), this also moves theme variables from `variables.scss` to `light.scss`, which always gets loaded anyways.

To test, compare the [deploy preview's kitchen sink Python code](https://deploy-preview-1166--just-the-docs.netlify.app/docs/index-test/#more-code) to the current [`main` branch's code](https://just-the-docs.github.io/just-the-docs/docs/index-test/#more-code); you can also use the "Preview dark color scheme" button to see OneDarkJekyll in action.

Users can opt-in to the old theme with `legacy_light`. I've documented this in the "customization" page.

Closes #679.

## implementation

Feel free to skip this part.

To do this, I:

- forked [mgyongyosi/OneDarkJekyll](https://github.com/mgyongyosi/OneDarkJekyll) to our own organization, [OneLightJekyll](https://github.com/just-the-docs/OneLightJekyll)
- updated the code to be slightly more robust (e.g. not require installing to global path)
- replaced the `colors.less` with the one pulled from Atom's [one-light-syntax](https://github.com/atom/atom/tree/master/packages/one-light-syntax)
- updated the license notice to also include GitHub's work in Atom
- regenerated the file
- plopped it in our current theme code



## next steps

This is related to #1100. I wanted to make this PR easier to review, so I won't implement that just yet; once we merge this, I can push that PR through. It's also related to #1173; in a comment below, I've also outlined some potential future work.

After we merge this, I'll trigger a release soon. I think this next minor bump can be focused on color schemes, e.g. dark theme, theme switching, and some bugfixes.

mattxwang commented:
Okay, I think this PR is ready for review. The only thing left to do once we start discussion on this PR is to document this change somewhere; I think UI Components > Code can use a new subsubheading describing the various highlight themes.

julianfortune commented:
The new color schemes look awesome to me!

pdmosses commented:
@mattxwang Thanks for this PR! The new code highlighting is a significant improvement.

> * forked [mgyongyosi/OneDarkJekyll](mgyongyosi/OneDarkJekyll) to our own organization, [OneLightJekyll](https://github.com/just-the-docs/OneLightJekyll)

The link to OneDarkJekyll is broken, it should be [mgyongyosi/OneDarkJekyll](https://github.com/mgyongyosi/OneDarkJekyll).

> To test, compare the [deploy preview's kitchen sink Python code](https://deploy-preview-1166--just-the-docs.netlify.app/docs/index-test/#more-code) to the current [`main` branch's code](https://just-the-docs.github.io/just-the-docs/docs/index-test/#more-code).

Perhaps point out that there's a button at the top of the kitchen sink to preview OneDarkJekyll.

mattxwang commented:
Thanks @pdmosses, just updated the PR description!

pdmosses commented:
> This is related to #1100. I wanted to make this PR easier to review, so I won't implement that just yet; once we merge this, I can push that PR through.

[My review](https://github.com/just-the-docs/just-the-docs/pull/1100#discussion_r1064049884) of #1100 isn't finished. That PR may need further work and discussion before approval.

mattxwang commented:
> [My review](https://github.com/just-the-docs/just-the-docs/pull/1100#discussion_r1064049884) of #1100 isn't finished. That PR may need further work and discussion before approval.

Ah, I could have been more clear. @simonebortolin (the original author) mentioned that he likely wouldn't have time to help out with more implementation work. What I'll do is make a new PR that solves the same issues in spirit (but split up some of the work into bite-sized PRs - like #1167).

agersant commented:
This new theme looks so much more readable 🥳

Thank you for working on this!

simonebortolin commented:
Yes in March I will be available

pdmosses commented on _sass/vendor/OneLightJekyll/syntax.scss:
```diff
@@ -0,0 +1,3 @@
+// pulls https://github.com/just-the-docs/OneLightJekyll, and changes file ext to .scss
+
+.highlight,pre.highlight{background:#f9f9f9;color:#383942}.highlight pre{background:#f9f9f9}.highlight .hll{background:#f9f9f9}.highlight .c{color:#9fa0a6;font-style:italic}.highlight .err{color:#fff;background-color:#e05151}.highlight .k{color:#a625a4}.highlight .l{color:#50a04f}.highlight .n{color:#383942}.highlight .o{color:#383942}.highlight .p{color:#383942}.highlight .cm{color:#9fa0a6;font- ...
```
I realise that this file was generated, and one shouldn't edit generated files. However, we also want our codebase to be readable by humans as well as computers, and the absence of layout significantly reduces its readability. Moreover, GitHub fails to add highlighting – compare with _sass/color_schemes/legacy_light.scss.

I know very little about how our CSS is checked during CI, but my recollection from making some minor changes (a couple of years ago) is that the checks were particularly fussy about layout. Has CI stopped checking CSS layout altogether? I don't understand how this file could ever pass the checks for 'prettier' layout.

mattxwang replied:
Great question! Files in `vendor/` are ignored by Stylelint:

https://github.com/just-the-docs/just-the-docs/blob/00cb3ba94b649c236163cfcf15c57ec684f576a5/.stylelintrc.json#L6

Stylelint is the only command on CI that runs with `npm test`. However, I can run prettier on the file to space it better and make it more readable. Just did that in https://github.com/just-the-docs/just-the-docs/pull/1166/commits/18e9dd93b715a80bfe9861c905448cd793e60710.

Separately, I can fix this inconsistency (right now, `npm run format` doesn't actually work as intended!)

mattxwang replied:
See: #1172 for a potential fix!

pdmosses reviewed (commented):
LICENSE-one-dark-jekyll surely shouldn't be included in _this_ PR! I know very little about licenses, but it looks like a mistake.

Ideally, the files in OneDarkJekyll and OneLightJekyll would have matching names (apart from the replacement of "dark" by "light"). So if we support the variant syntax-one-dark-vivid in OneDarkJekyll, I'd expect to find syntax-one-light-vivid in OneLightJekyll. Then when some website has a dark scheme using syntax-one-dark-vivid, they can make a _corresponding_ light scheme using syntax-one-light-vivid.

I'm confused by the need for both LICENSE and LICENSE-one-light, because syntax.scss was generated by a compiler from a copied file. Due to my lack of expertise, I shouldn't approve a PR that affects licenses: it needs a reviewer who would notice any problems.

pdmosses reviewed (commented):
Thanks for adding the layout!

mattxwang commented:
> Ideally, the files in OneDarkJekyll and OneLightJekyll would have matching names (apart from the replacement of "dark" by "light"). So if we support the variant syntax-one-dark-vivid in OneDarkJekyll, I'd expect to find syntax-one-light-vivid in OneLightJekyll. Then when some website has a dark scheme using syntax-one-dark-vivid, they can make a _corresponding_ light scheme using syntax-one-light-vivid.

I agree with this! I see that we added the vivid/firewatch themes in #578. Do you mind reminding me of the intention of the alternate themes? If there's a reasonable motivation, I'm happy to generate `one-light-*` SCSS files using our fork.

In particular, I see two paths:

1. Remove the vivid/firewatch syntax files, maintain only one "canonical" highlight theme
2. Add the equivalent files for light theme vivid/firewatch highlighting

> LICENSE-one-dark-jekyll surely shouldn't be included in _this_ PR! I know very little about licenses, but it looks like a mistake.
> 
> I'm confused by the need for both LICENSE and LICENSE-one-light, because syntax.scss was generated by a compiler from a copied file. Due to my lack of expertise, I shouldn't approve a PR that affects licenses: it needs a reviewer who would notice any problems.

This is a good point, and I'm also not a lawyer either. My reasoning for including the licenses is that, in my opinion, the OneLightJekyll code is **so derivative** from the two other projects, that this clause in the two MIT licenses:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

applies, i.e. I think this is a "substantial portion" of both the one-light Atom theme and the OneDarkJekyll codebase. So, when we distribute the source code for OneLightJekyll, we should also distribute the relevant licenses.

In contrast, post-compilation none of the license files make their way into the `_site` build, so adding these LICENSE files doesn't affect the end user. In addition, I think at that point, the work is no longer derivative in the same way.

pdmosses commented:
> I see that we added the vivid/firewatch themes in #578. Do you mind reminding me of the intention of the alternate themes?

There's some discussion around https://github.com/just-the-docs/just-the-docs/pull/578#issuecomment-1170240433, but it doesn't consider whether OneDarkJekyll naturally includes those variants. I think I included them simply because they were there, and I was aiming to make a straightforward copy.

pdmosses commented:
> > LICENSE-one-dark-jekyll surely shouldn't be included in _this_ PR! I know very little about licenses, but it looks like a mistake.

That point wasn't about the details of licenses: it was simply that the file belongs in One***Dark***Jekyll, rather than in One***Light***Jekyll.

mattxwang commented:
> That point wasn't about the details of licenses: it was simply that the file belongs in One***Dark***Jekyll, rather than in One***Light***Jekyll.

Oh, I see - I misunderstood your original comment, sorry!

I think the file is relevant for both:

- we need to include the license for OneDarkJekyll (this already exists)
- *separately*, we need to include *all* the relevant licenses for OneLightJekyll. Per the previous comment, OneLightJekyll is derivative of OneDarkJekyll, and thus I've included the license (and the one for the one-light theme)

What are your thoughts?

>  I think I included them simply because they were there, and I was aiming to make a straightforward copy.

Gotcha; if there's no strong rationale, then I'm leaning towards deprecating + removing them in the minor patch *after* we make this change. It's more to support, and I think users looking to change their highlight theme can do so by making a completely new custom theme.

With that in mind, I'm thinking that we shouldn't add the corresponding light themes. What do you think?

pdmosses commented:
> I think users looking to change their highlight theme can do so by making a completely new custom theme.

Right: if they want to use One Dark Vivid, they might also want to adjust their main scheme colours.

Should users simply add files such as `_sass/vendor/OneDarkJekyll/syntax-one-dark-vivid.scss`, and rely on the theme having provided `LICENSE-one-dark-jekyll`, or do they need to add a local copy of the license? Either way, it would be helpful to mention it in the docs.

>  I'm thinking that we shouldn't add the corresponding light themes. What do you think?

I agree. One Dark Vivid and Firewatch are merely _examples_ of Atom themes. https://github.com/mgyongyosi/OneDarkJekyll states:

> t's LESS file which can be compiled to a - Pygments, Rouge compatible - stylesheet from Atom editor's One Dark syntax theme (and any theme based on it, for example One Dark Vivid, Firewatch, etc.).

We clearly don't want to include all themes based on One Dark, so including just the main One Dark theme seems sensible. The OneDarkJekyll repo might add more examples, but we shouldn't feel obliged to add them to JtD.

mattxwang commented:
> Should users simply add files such as `_sass/vendor/OneDarkJekyll/syntax-one-dark-vivid.scss`, and rely on the theme having provided LICENSE-one-dark-jekyll, or do they need to add a local copy of the license?

My understanding is that if they distribute the source code at all (e.g. put it on GitHub), they should include their own license file for each addition. However, I don't think it's our obligation as a theme to do this; in general, I haven't seen open-source software advise users on specific license issues (I think that puts us in a strange liability spot).

> We clearly don't want to include all themes based on One Dark, so including just the main One Dark theme seems sensible. The OneDarkJekyll repo might add more examples, but we shouldn't feel obliged to add them to JtD. Similarly for One Light.

Sounds good. For this PR, what do you think are the next steps (assuming that we aren't adding anymore One Light themes)?

Separately, I can open up an issue / PR to remove the (undocumented) extra One Dark files, or figure out an alternative solution.

mattxwang commented:
Just renamed / moved the relevant files so the directories are identical!

I'm a bit wary of moving the variables since technically this changes code behaviour (ex now `setup.scss` doesn't override those variables). What do you think? LI can also sleep on it and investigate ways to resolve that.

pdmosses commented on _sass/color_schemes/dark.scss:
```diff
@@ -12,21 +12,7 @@ $search-background-color: $grey-dk-250;
-// The following highlight theme is more legible than that used for the light color scheme
-
-// @import "./vendor/OneDarkJekyll/syntax-one-dark";
-// $code-background-color: #282c34; // OneDarkJekyll default for syntax-one-dark
-// $code-linenumber-color: #abb2bf; // OneDarkJekyll .nf for syntax-one-dark
```
Shouldn't these code colors be used instead of those for syntax-one-dark-vivid?

pdmosses commented on _sass/color_schemes/dark.scss:
```diff
@@ -12,21 +12,7 @@ $search-background-color: $grey-dk-250;
-// $code-linenumber-color: #abb2bf; // OneDarkJekyll .nf for syntax-one-dark
-
-@import "./vendor/OneDarkJekyll/syntax-one-dark-vivid";
+@import "./vendor/OneDarkJekyll/syntax";
```
I suggest to comment here that syntax.scss is actually a copy of OneDarkJekyll/syntax-one-dark-vivid.scss. The comments referring to syntax-one-dark-vivid on the code color rules confused me.

pdmosses commented:
> I'm a bit wary of moving the variables since technically this changes code behaviour (ex now `setup.scss` doesn't override those variables). What do you think? LI can also sleep on it and investigate ways to resolve that.

I think it's _essential_ to move them! 

Currently, `setup.scss` doesn't override the variables set in `_sass/color_schemes/dark.scss` (nor in any color scheme other than `light`). If a toggle or media-prefers changes between the `light` and `dark` color schemes, it would surely be very confusing if it affects whether the variables in `setup.scss` are overriding or not.

Moreover, all the variable settings in question are marked `!default`. As you [noted in another PR](https://github.com/just-the-docs/just-the-docs/pull/1173#user-content-fn-1-1f6bdc0cb25bf85a733c828ed744023b), that _ought_ to avoid overriding variables that have already been set, but doesn't…

To make `dark.scss` analogous to `light.scss`, I think its settings should also be `!default`s.

mattxwang commented:
Thanks for your patience, I've given it some more thought. I don't want to barrage you with too much content, so have split some other observations into subheadings; they're not strictly necessary for this PR. The tl;dr is that I've implemented the suggested change (and ordered the variables slightly).

> Note however that $code-background-color should perhaps be copied from the highlighting background color, instead of being set independently of the highlighting used.

Let's make this a bite-sized separate change, I think it's a slightly separate purpose / bug fix.

### ordering of `setup.scss`

Separately, I think you've pointed out a valid point of confusion / docs mistake from when I implemented `setup.scss`.

In particular, the [docs for `setup.scss`](https://just-the-docs.com/docs/customization/#override-and-define-new-variables) reads:

> To define new SCSS variables, functions, or override existing theme variables, place SCSS code in `_sass/custom/setup.scss`

But, this is not true - `setup.scss` is loaded *before* all of the themes, so it doesn't override existing theme variables. 

In my opinion, the solution here is to move `setup.scss` after all the `color_scheme` SCSS. This way, `setup.scss` properly allows theme overrides. In addition, users who previously defined variables in `setup.scss` and then used them in their custom color schemes can shift those declarations to be entirely in the theme code.

To me, that resolves the problem that you've brought up without requiring a variable move.

### `!default`

Separately,

> To make dark.scss analogous to light.scss, I think its settings should also be !defaults.

Since we're not using `!default` directly (i.e., it doesn't work), I don't think we should be adding any more `!default`s - it's misleading, since users may think that the annotation works. In the long-term, I think the order of operations should be:

1. First, fix our use of `!default`; this requires:
    - us to restructure how custom schemes work in a non-trivial breaking way
    - possibly, require newer SCSS features in `dart-sass`, which is not used by our users who use the `github-pages` gem
2. Then, properly annotate variables with `!default` - I think this is just all of our variables, but I'll have to do a deeper drive
3. Then, properly document *how to use default variables*, which is not what the average non-technical user would think

What are your thoughts?

mattxwang commented:
Opened #1184 to address one of the issues I brought up!

pdmosses commented on _sass/vendor/OneLightJekyll/syntax.scss:
```diff
@@ -1,51 +1,53 @@
+// Generated with OneLightJekyll applied to Atom's One Light theme
```
The diff claims:
```
...OneDarkJekyll/syntax-firewatch-green.scss → _sass/vendor/OneLightJekyll/syntax.scss
```
That seems incompatible with the comment. I haven't tried check the color codes.

mattxwang replied:
Gotcha. The comment is correct; I don't manually control the git diff and how it chooses to add/rename files (my understanding is that GitHub uses the default git engine to determine the "files changed" tab). Let me see if I can do some hacky commits to override this history.

mattxwang replied:
Okay, after doing some local playing around, it'll be *very* hard to simultaneously remove the old dark themes + create this new one, without git marking it as a rename-and-edit; the contents of the file are too similar.

I'd prefer to make this simpler; I think we can just merge this PR first (squashing the commits, etc.) and then immediately follow-up with another PR to remove the other syntax themes. What are your thoughts?

pdmosses replied:
LGTM!

pdmosses reviewed (changes requested):
Thanks for implementing this! I think the uniformity between light and dark will be especially helpful when toggling is added.

Before merging, please resolve the apparent incompatibility between the comment at the top of OneLightJekyll/syntax.scss and the move that created the file, which confused me.
